File: 239. 滑动窗口最大值.py
from collections import deque


class Solution(object):
    def maxSlidingWindow(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        result = []
        q = deque()
        for index, num in enumerate(nums):
            # 1. 劝退：保持单调递减
            while q and nums[q[-1]] <= num:
                q.pop()
            q.append(index)
            # 2. 过期：检查队首下标是否已经超出了窗口范围
            # 窗口范围是 [index-k+1, index]，所以下标必须 > index-k
            if q[0]==index - k:
                q.popleft()
            # 3. 记录：只要窗口形成了，队首就是当前窗口的最大值
            if index >= k - 1:
                result.append(nums[q[0]])
        return result

File: 239. 滑动窗口最大值.go
package main

func maxSlidingWindow(nums []int, k int) []int {
	res := make([]int, 0, len(nums)-k+1)
	mayBeMax := make([]int, 0)

	// 直接一个循环搞定，不需要显式维护 l 和 r
	for i := range len(nums) {

		// 只要切片不为空，且当前新人 nums[i] >= 尾部对应的值，就无情踢掉尾部
		for len(mayBeMax) > 0 && nums[i] >= nums[mayBeMax[len(mayBeMax)-1]] {
			mayBeMax = mayBeMax[:len(mayBeMax)-1]
		}

		mayBeMax = append(mayBeMax, i)

		// 清理过期数据
		if mayBeMax[0] < i-k+1 {
			mayBeMax = mayBeMax[1:]
		}

		if i >= k-1 {
			res = append(res, nums[mayBeMax[0]])
		}
	}

	return res
}
